### PR TITLE
[Parked] Fixes #263: Handle network request timeouts.

### LINKED ISSUE #263: Handle network request timeouts.
Right now when a server is down/unreachable, a first `OPTIONS` request sent will hang forever. 

We need to support request timeouts with a configurable value. Possible pattern to use:

```js
var timeoutPromise = new Promise(function(resolve) {
  setTimeout(resolve, 10000);
})

var load = fetch(‘/some-resource’);

Promise.race([timeoutPromise, load]).then(function(res) {
  if (res instanceof Response) {
    // Network replied in time.
  } else {
    // Timeout
  }
})
```

### PR DESCRIPTION
**Pull request by n1k0**
refs #263. r=? @leplatrem

### CODE REVIEW

**Comment by n1k0** [comment]
Also cc @michielbdejong for feedback

**Comment by Natim** [comment]
LGTM.

File: src/http.js [hunk 1]
```diff
@@ -108,6 +131,11 @@ export default class HTTP {
         }
         return {status, json, headers};
       });
+    return Promise.race([timeoutPromise, requestPromise])
```
Comment (leplatrem) [inline src/http.js]: nit: add a small comment to specify that the quickest one is resolved

File: src/http.js [hunk 2]
```diff
@@ -62,11 +80,16 @@ export default class HTTP {
...
    */
   request(url, options={headers:{}}) {
-    let response, status, statusText, headers;
+    let response, status, statusText, headers, _timeout;
     // Ensure default request headers are always set
     options.headers = Object.assign({}, HTTP.DEFAULT_REQUEST_HEADERS, options.headers);
     options.mode = this.requestMode;
-    return fetch(url, options)
+    const timeoutPromise = new Promise((_, reject) => {
+      _timeout = setTimeout(() => {
```
Comment (leplatrem) [inline src/http.js]: `_timeoutId` ? (*not sure if relevant*)

**Comment by leplatrem** [comment]
r+!

**Comment by n1k0** [comment]
I'm realizing there's no way to guarantee the request is aborted once the timeout value is reached, see https://github.com/whatwg/fetch/issues/27.

That's annoying, as the user may believe the request was never actually performed, while it totally may, resulting in a possibly messy situation. 

I don't have any solution in mind right now.

**Comment by n1k0** [comment]
Issue is partially confirmed by changing the test like this:

```js
    describe("Request timeout", () => {
      it("should timeout the request", () => {
        sandbox.stub(root, "fetch").returns(
          new Promise(resolve => {
            console.log("executed :(");
            setTimeout(resolve, 20000);
          }));
        return http.request("/")
          .should.eventually.be.rejectedWith(Error, /timeout/);
      });
    });
``` 

Where "executed :(" is printed out to the console regardless.

**Comment by n1k0** [comment]
The only option to properly support this would be to switch back to XHR, which supports timeouts natively. I'm just gonna park this for now out of depression ;)

**Comment by michielbdejong** [comment]
Ah, could well be the fix for https://bugzilla.mozilla.org/show_bug.cgi?id=1224203 :)

**Comment by michielbdejong** [comment]
Can't we just set a variable 'hasTimedOut' and check for this before returning success?

**Comment by n1k0** [comment]
Fixed by #273

**Pull request closed without merging**